File: 2999_count_the_number_of_powerful_integers.java
class Solution {
    long[] t = new long[17], p = new long[17];

    void z() {
        if (t[0] == 1) return;
        t[0] = 1;
        for (int i = 1; i < 17; i++) t[i] = t[i - 1] * 10;
    }

    void y(int b) {
        p[0] = 1;
        for (int i = 1; i < 17; i++) p[i] = p[i - 1] * b;
    }

    long x(long u, long v, int d, int m) {
        if (u < v) return 0;
        long u0 = u / t[d], r = u % t[d], z = 0;
        int j = 0;
        while (j < 17 && t[j] <= u0) j++;
        for (int i = j - 1; i >= 0; i--) {
            long q = u0 / t[i];
            if (q > m) return z + p[i + 1];
            z += p[i] * q;
            u0 -= q * t[i];
        }
        return z + (r >= v ? 1 : 0);
    }

    public long numberOfPowerfulInt(long a, long b, int m, String s) {
        z();
        y(m + 1);
        long v = Long.parseLong(s);
        return x(b, v, s.length(), m) - x(a - 1, v, s.length(), m);
    }
}

File: 2999_count_the_number_of_powerful_integers.cpp
long long u[17], v[17];
class Solution {
public:
    void r() { u[0] = 1; for (int i = 1; i < 17; i++) u[i] = u[i - 1] * 10; }
    void s(int a) { v[0] = 1; for (int i = 1; i < 17; i++) v[i] = v[i - 1] * a; }
    long long t(long long x, long long y, int z, int k) {
        if (x < y) return 0;
        long long p = x / u[z], q = x % u[z], ans = 0;
        int i = lower_bound(u, u + 17, p) - u;
        while (i >= 0) {
            auto [a, b] = ldiv(p, u[i]);
            if (a > k) return ans + v[i + 1];
            ans += v[i] * a;
            p -= a * u[i--];
        }
        return ans + (q >= y);
    }
    long long numberOfPowerfulInt(long long a, long long b, int c, string& d) {
        r(); s(c + 1);
        return t(b, stoll(d), d.size(), c) - t(a - 1, stoll(d), d.size(), c);
    }
};
